### PR TITLE
added notifier

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/1616.

### CODE REVIEW

File: flag/service/notifier/notifier.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package notifier
+
+import (
+	"github.com/giantswarm/draughtsman-operator/flag/service/notifier/slack"
+)
+
+type Notifier struct {
```
Comment (xh3b4sd) [inline flag/service/notifier/notifier.go]: This is basically the same as in draughtsman.

File: service/installer/error.go [hunk 2]
```diff
@@ -1,12 +1,10 @@
 package installer
 
-import (
-	"github.com/juju/errgo"
-)
+import "github.com/giantswarm/microerror"
```
Comment (xh3b4sd) [inline service/installer/error.go]: Boyscouting here.

File: service/notifier/notifier.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		slackConfig := slacknotifier.DefaultConfig()
+
+		slackConfig.Logger = config.Logger
+		slackConfig.SlackClient = slack.New(config.Viper.GetString(config.Flag.Service.Notifier.Slack.Token))
+
+		slackConfig.Channel = config.Viper.GetString(config.Flag.Service.Notifier.Slack.Channel)
+		slackConfig.Emoji = config.Viper.GetString(config.Flag.Service.Notifier.Slack.Emoji)
+		// TODO configure env in separate PR when adding eventer to sort out deployment event status updates.
+		//slackConfig.Environment = config.Viper.GetString(config.Flag.Service.Eventer.Environment)
+		slackConfig.Environment = "TODO"
```
Comment (xh3b4sd) [inline service/notifier/notifier.go]: This is my next step. To make this work I realized we need to extend the TPR project with the deployment event ID.

File: service/notifier/slack/slack.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		return nil, microerror.Maskf(invalidConfigError, "config.Emoji must not be empty")
+	}
+	if config.Environment == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Environment must not be empty")
+	}
+	if config.Username == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.Username must not be empty")
+	}
+
+	// TODO add common health check for this.
```
Comment (xh3b4sd) [inline service/notifier/slack/slack.go]: Something for later.

File: service/notifier/slack/slack.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	attachment.MarkdownIn = []string{"text"}
+
+	attachment.Title = fmt.Sprintf(titleFormat, project.Name, project.Ref)
+	attachment.Text = fmt.Sprintf(failedMessageFormat, errorMessage)
+	if success {
+		attachment.Text = successMessage
+	}
+	attachment.Footer = n.environment
```
Comment (xh3b4sd) [inline service/notifier/slack/slack.go]: The footer misses the deployment event ID. We do not have it yet. We can get this straight in the next step when we add the deployment event ID to the TPR.

File: service/resource/project/resource.go [hunk 6]
```diff
@@ -185,9 +193,19 @@ func (r *Resource) ProcessCreateState(obj, createState interface{}) error {
...
+				if notiErr != nil {
+					return microerror.Mask(notiErr)
+				}
+
+				return microerror.Mask(instErr)
+			}
+
+			notiErr := r.notifier.Success(notifierspec.Project{Name: p.Name, Ref: p.Ref})
+			if notiErr != nil {
+				return microerror.Mask(notiErr)
```
Comment (xh3b4sd) [inline service/resource/project/resource.go]: This code is duplicated below. Good enough for me now. Before I improve this somehow I would rather like to improve the difference between installing and upgrading charts via helm.

File: main.go [hunk 7]
```diff
@@ -157,6 +158,12 @@ func mainWithError() error {
 	daemonCommand.PersistentFlags().String(f.Service.Configurer.Secret.Namespace, "draughtsman", "Namespace of secret holding values data.")
 	daemonCommand.PersistentFlags().String(f.Service.Configurer.Types, string(configmap.ConfigurerType)+","+string(secret.ConfigurerType), "Comma separated list of configurers to use for configuration management.")
 
+	daemonCommand.PersistentFlags().String(f.Service.Notifier.Slack.Channel, "", "Channel to post Slack notifications to.")
+	daemonCommand.PersistentFlags().String(f.Service.Notifier.Slack.Emoji, ":older_man:", "Emoji to use for Slack notifications.")
```
Comment (rossf7) [inline main.go]: <3

**Pull request merged**